docs/src/consensus/transaction.md

```diff
@@ -0,0 +1,178 @@
+   - set `kernel.coinbase` to `txa.coinbase` or `txb.coinbase` or to `None`
+   - set `kernel.fee` to `txa.fee + txb.fee`
+   - set `kernel.timestamp` to `max(txa.timestamp, txb.timestamp)`
+   - set `kernel.mutator_set_hash` to `txa.mutator_set_hash`
+ - compute the mast hash of `kernel`
+ - verify the computed hash against the given `transaction_kernel_mast_hash`.
+
+### E: Proof of Integral Mempool Operation
+
+A transaction is valid if it was ever added to an integral mempool. The motivating use case for this feature is that mempool operators can delete transaction proofs as long as they store and routinely update one 
+
+An integral mempool is an MMR containing transactions *kernels*, along with a proof of integral history. The integral mempool can be updated in only one way: by appending a valid transaction.
+ - `append : (old_mmr : Mmr<TransactionKernel>) × (old_history_proof: StarkProof) × (tx : Transaction) ⟶ (new_mmr : Mmr<TransactionKernel>) × (new_history_proof : StarkProof)`
+
+The proof of integral history certifies that the MMR is the (left-hand side of the) output of some `append` operation. Specifically, the claim is the input-output-program triple
+ - input: `mmr : Mmr<TransactionKernel>`
+ - output: `∅`
+ - program:
+   - if `mmr` is empty, halt gracefully; otherwise
+   - divine `old_mmr`
+   - divine `tx_kernel`
+   - verify `tx_kernel` with some divined transaction proof
+   - append `tx_kernel` to `old_mmr` resulting in `new_mmr`
+   - assert that `new_mmr == mmr`.
+
+The claim for certifying the validity of transaction based on its inclusion in an integral mempool is induced by `MemberOfIntegralMempool :: (transaction_kernel_mast_hash : Digest) ⟶ ∅`, and the program executes the following logic:
+ - divine `mmr : Mmr<TransactionKernel>`
+ - verify `mmr` with some divined proof of integral history
+ - verify membership of `transaction_kernel_mast_hash` to `mmr` with a divined authentication path.
+
+### F: Transaction Data Update
+
+A transaction is valid if another transaction that is identical except for fixing an older mutator set hash or timestamp, was valid. Specifically, the program The program `TransactionDataUpdate :: (transaction_kernel_mast_hash : Digest) ⟶ ∅` verifies the update of transaction data as follows:
+ - divine `tx : TransactionKernel`
+ - verify `tx` with some divined proof
+ - create a new `TransactionKernel` object `kernel`
+ - set all fields of `kernel` to the matching field of `tx` except:
+   - set `kernel.timestamp` such that `kernel.timestamp >= tx.timestamp`
+   - set `kernel.mutator_set_hash =/= tx.mutator_set_hash` only if the following instructions execute gracefully without crashing
+     - divine the mutator set AOCL MMR accumulator `kernel_aocl`
+     - authenticate `kernel_aocl` against the mutator set MAST hash `kernel.mutator_set_hash` using a divined authentication path
```

Comment:
Do we also need to deal with the removal records here?

Reply:
The removal records contain authentication paths into the SWBF MMR. Those authentication paths need to be updated whenever the tip is changed.